File: github.go
package main

import (
	"bytes"
	"crypto"
	"crypto/rand"
	"crypto/rsa"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"math"
	"mime"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"
)

const (
	DefaultGitHubAPI     GitHubAPI     = "https://api.github.com"
	DefaultGitHubGraphQL GitHubGraphQL = "https://api.github.com/graphql"
)

type GitHubToken string

type GitHubAPI string

func (gh GitHubAPI) AppJWT(appID int64, key *rsa.PrivateKey) (GitHubToken, error) {
	var jwt []byte
	jwt = base64.RawURLEncoding.AppendEncode(jwt, []byte(`{"alg":"RS256","typ":"JWT"}`))
	jwt = append(jwt, '.')
	jwt = base64.RawURLEncoding.AppendEncode(jwt, []byte(`{`+
		`"iat":`+strconv.FormatInt(time.Now().Add(-time.Minute).Unix(), 10)+
		`,"exp":`+strconv.FormatInt(time.Now().Add(time.Minute*1).Unix(), 10)+
		`,"iss":"`+strconv.FormatInt(appID, 10)+`"}`))

	sha := sha256.Sum256(jwt)
	sig, err := rsa.SignPKCS1v15(rand.Reader, key, crypto.SHA256, sha[:])
	if err != nil {
		return "", err
	}

	jwt = append(jwt, '.')
	jwt = base64.RawURLEncoding.AppendEncode(jwt, sig)
	return GitHubToken(jwt), nil
}

func (gh GitHubAPI) GetRepoInstallation(jwt GitHubToken, repo string) (int64, error) {
	verbose("getting app installation id for repo %q", repo)
	resp, buf, err := gh.request(jwt, http.MethodGet, "repos/"+repo+"/installation", nil)
	if err != nil {
		return 0, err
	}
	if resp.StatusCode != http.StatusOK {
		return 0, fmt.Errorf("response status %d (body: %q)", resp.StatusCode, buf)
	}

	var obj struct {
		ID int64 `json:"id"`
	}
	if err := json.Unmarshal(buf, &obj); err != nil {
		return 0, fmt.Errorf("parse response: %w", err)
	}
	if obj.ID == 0 {
		return 0, fmt.Errorf("parse response: missing installation id")
	}
	verbose("got installation id %d", obj.ID)

	return obj.ID, nil
}

func (gh GitHubAPI) CreateInstallationToken(jwt GitHubToken, repo string, installID int64) (GitHubToken, error) {
	if _, r, ok := strings.Cut(repo, "/"); ok {
		repo = r
	}

	verbose("getting app installation token for repo %q", repo)
	resp, buf, err := gh.request(jwt, http.MethodPost, "app/installations/"+strconv.FormatInt(installID, 10)+"/access_tokens", map[string]any{
		"repositories": []string{repo},
		"permissions": map[string]string{
			"contents": "write",
		},
	})
	if err != nil {
		return "", err
	}
	if resp.StatusCode != http.StatusCreated {
		return "", fmt.Errorf("response status %d (body: %q)", resp.StatusCode, buf)
	}

	var obj struct {
		Token       string `json:"token"`
		Permissions struct {
			Contents string `json:"contents"`
		} `json:"permissions"`
	}
	if err := json.Unmarshal(buf, &obj); err != nil {
		return "", fmt.Errorf("parse response: %w", err)
	}
	if obj.Token == "" {
		return "", fmt.Errorf("parse response: missing token")
	}
	if obj.Permissions.Contents != "write" {
		return "", fmt.Errorf("installation does not have contents:write access")
	}
	verbose("got installation token")

	return GitHubToken(obj.Token), nil
}

func (gh GitHubAPI) RevokeInstallationToken(token GitHubToken) error {
	verbose("revoking app installation token")
	resp, buf, err := gh.request(token, http.MethodDelete, "installation/token", nil)
	if err != nil {
		return err
	}
	if resp.StatusCode != http.StatusNoContent {
		return fmt.Errorf("response status %d (body: %q)", resp.StatusCode, buf)
	}
	verbose("app installation token revoked")

	return nil
}

func (gh GitHubAPI) request(token GitHubToken, method, path string, body any) (*http.Response, []byte, error) {
	u, err := url.Parse(string(gh))
	if err != nil {
		return nil, nil, fmt.Errorf("create request: %w", err)
	}
	u = u.JoinPath("/", path) // not ResolveReference since we just want to append the elements

	var (
		r        io.Reader
		bodyJSON []byte
	)
	if body != nil {
		bodyJSON, err = json.Marshal(body)
		if err != nil {
			return nil, nil, fmt.Errorf("marshal request: %w", err)
		}
		r = bytes.NewReader(bodyJSON)
	}
	req, err := http.NewRequest(method, u.String(), r)
	if err != nil {
		return nil, nil, fmt.Errorf("create request: %w", err)
	}
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("Content-Length", strconv.Itoa(len(bodyJSON)))
	}

	req.Header.Set("Accept", "application/vnd.github+json")
	req.Header.Set("X-GitHub-Api-Version", "2026-03-10")

	if token != "" {
		req.Header.Set("Authorization", "Bearer "+string(token))
	}
	if UserAgent != "" {
		req.Header.Set("User-Agent", UserAgent)
	}

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, nil, fmt.Errorf("do request: %w", err)
	}
	defer resp.Body.Close()

	buf, err := io.ReadAll(resp.Body)
	if err != nil {
		err = fmt.Errorf("read response: %w", err)
	}
	return resp, buf, err
}

type CreateCommitOnBranchInput struct {
	Branch          CommittableBranch `json:"branch"`
	ExpectedHeadOid GitObjectID       `json:"expectedHeadOid"`
	Message         CommitMessage     `json:"message"`
	FileChanges     FileChanges       `json:"fileChanges"`
}

type CommittableBranch struct {
	RepositoryNameWithOwner string `json:"repositoryNameWithOwner"`
	BranchName              string `json:"branchName"`
}

type GitObjectID = OID

type CommitMessage struct {
	Headline string `json:"headline"`
	Body     string `json:"body"`
}

type FileChanges struct {
	Additions []FileAddition `json:"additions"`
	Deletions []FileDeletion `json:"deletions"`
}

type FileAddition struct {
	Contents Base64String `json:"contents"`
	Path     string       `json:"path"`
}

type FileDeletion struct {
	Path string `json:"path"`
}

type Base64String []byte

var _ json.Marshaler = (Base64String)(nil)
var _ json.Unmarshaler = (*Base64String)(nil)

func (b Base64String) MarshalJSON() ([]byte, error) {
	dst := make([]byte, 0, base64.StdEncoding.EncodedLen(len(b))+2)
	dst = append(dst, '"')
	dst = base64.StdEncoding.AppendEncode(dst, b)
	dst = append(dst, '"')
	return dst, nil
}

func (b *Base64String) UnmarshalJSON(buf []byte) error {
	var s string
	if err := json.Unmarshal(buf, &s); err != nil {
		return err
	}
	if x, err := base64.StdEncoding.DecodeString(s); err != nil {
		return err
	} else {
		*b = Base64String(x)
	}
	return nil
}

type GitHubGraphQL string

func (gh GitHubGraphQL) CreateCommitOnBranch(token GitHubToken, input CreateCommitOnBranchInput) (OID, error) {
	query := `
		mutation($input: CreateCommitOnBranchInput!) {
			createCommitOnBranch(input: $input) {
				commit {
					oid
				}
			}
		}
	`
	data, err := gh.query(token, query, map[string]any{"input": input})
	if err != nil {
		// best-effort attempt to have better errors for certain cases (as of 2026-04-04)
		switch {
		case strings.Contains(err.Error(), "No commit exists with specified expectedHeadOid"):
			return "", fmt.Errorf("remote branch head is behind local parent commit (error: %w)", err)
		case strings.Contains(err.Error(), "Expected branch to point to"):
			return "", fmt.Errorf("local parent commit is behind remote branch head (error: %w)", err)
		}
		return "", err
	}
	var obj struct {
		CreateCommitOnBranch struct {
			Commit struct {
				OID OID `json:"oid"`
			} `json:"commit"`
		} `json:"createCommitOnBranch"`
	}
	if err := json.Unmarshal(data, &obj); err != nil {
		return "", fmt.Errorf("parse response: %w", err)
	}
	if !obj.CreateCommitOnBranch.Commit.OID.Valid() {
		return "", fmt.Errorf("github created the commit but returned an empty/invalid oid %q", obj.CreateCommitOnBranch.Commit.OID)
	}
	return obj.CreateCommitOnBranch.Commit.OID, nil
}

var ghMutationRateLimit *time.Ticker

func (gh GitHubGraphQL) query(token GitHubToken, query string, variables map[string]any) (json.RawMessage, error) {
	var reqObj struct {
		Query     string         `json:"query"`
		Variables map[string]any `json:"variables"`
	}
	reqObj.Query = query
	reqObj.Variables = variables

	reqObjJSON, err := json.Marshal(reqObj)
	if err != nil {
		return nil, fmt.Errorf("marshal request: %w", err)
	}
	verbose("request body size %d", len(reqObjJSON))

	req, err := http.NewRequest(http.MethodPost, string(gh), bytes.NewReader(reqObjJSON))
	if err != nil {
		return nil, fmt.Errorf("create request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Content-Length", strconv.Itoa(len(reqObjJSON)))
	req.Header.Set("Accept", "application/json")

	if token != "" {
		req.Header.Set("Authorization", "Bearer "+string(token))
	}
	if UserAgent != "" {
		req.Header.Set("User-Agent", UserAgent)
	}

	var buf []byte
	for try := 1; ; try++ {
		if ghMutationRateLimit == nil {
			ghMutationRateLimit = time.NewTicker(time.Second) // https://docs.github.com/en/graphql/overview/rate-limits-and-query-limits-for-the-graphql-api#staying-under-the-rate-limit
		} else {
			select {
			case <-ghMutationRateLimit.C:
			default:
				verbose("delaying request to stay within rate limit")
				<-ghMutationRateLimit.C
			}
		}

		resp, err := http.DefaultClient.Do(req)
		if err != nil {
			return nil, fmt.Errorf("do request: %w", err)
		}
		buf, err = io.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			return nil, fmt.Errorf("read response: %w", err)
		}

		verbose("response status %d", resp.StatusCode)
		for k, vs := range resp.Header {
			for _, v := range vs {
				verbose("response header %s = %q", k, v)
			}
		}
		verbose("response body %q", buf)

		// https://github.com/octokit/plugin-retry.js/blob/v8.1.0/src/index.ts
		if resp.StatusCode >= 400 {
			switch resp.StatusCode {
			case 400, 401, 403, 404, 410, 422, 451:
				return nil, fmt.Errorf("non-retryable response status %d (try: %d, body: %q)", try, resp.StatusCode, buf)
			}
			if try > 3 {
				return nil, fmt.Errorf("response status %d, no retries left (try: %d, body: %q)", try, resp.StatusCode, buf)
			}
			retryAfter := time.Duration(float64(time.Second) * math.Pow(float64(try), 2))
			warning("request failed, will retry after %s (response status %d, body %q)", retryAfter, resp.StatusCode, buf)
			<-time.After(retryAfter)
			continue
		}

		if mt, _, _ := mime.ParseMediaType(req.Header.Get("Content-Type")); mt != "application/json" {
			if resp.StatusCode != 200 {
				return nil, fmt.Errorf("response status %d (body: %q)", resp.StatusCode, buf)
			}
			return nil, fmt.Errorf("incorrect response type %q", mt)
		}
		break
	}

	var respObj struct {
		Errors []struct {
			Type    string `json:"type"`
			Message string `json:"message"`
		} `json:"errors"`
		Data json.RawMessage `json:"data"`
	}
	if err := json.Unmarshal(buf, &respObj); err != nil {
		return nil, fmt.Errorf("parse response: %w", err)
	}

	var errs []error
	for _, e := range respObj.Errors {
		errs = append(errs, fmt.Errorf("%s: %s", e.Type, e.Message))
	}
	if err := errors.Join(errs...); err != nil {
		return nil, fmt.Errorf("github failed to create commit: %w", err)
	}
	return respObj.Data, nil
}

File: lib/core/github.ts
import { type KeyObject, createSign } from 'node:crypto'
import type { OID } from './git.ts'
import { debuglog, jsonify } from '../util/util.ts'

const debug = debuglog('github') // NODE_DEBUG=github

/** A GitHub token. */
export type GitHubToken = string & { __token: true }
export type GitHubAppJwt = GitHubToken & { __token_appjwt: true }
export type GitHubInstallationToken = GitHubToken & { __token_installation: true }

/** The GitHub REST API base URL. */
export type GitHubApiUrl = string & { __ghapi: true }

/** The GitHub GraphQL API base URL. */
export type GitHubGraphqlUrl = string & { __ghgqlapi: true }

export const DefaultGitHubApi = 'https://api.github.com' as GitHubApiUrl
export const DefaultGitHubGraphql = 'https://api.github.com/graphql' as GitHubGraphqlUrl

let globalUserAgent = ''

/** Sets the user agent used for requests. */
export function setUserAgent(ua: string): void {
  globalUserAgent = ua
}

interface TokenProps {
  maxRetries: number,
  userAgent: string,
}

type TokenProp = keyof TokenProps
type TokenWithProp<T extends TokenProp> = GitHubToken & { [K in T]: TokenProps[T] }

/** Returns a copy of token with retries enabled. */
export function withRetries(token: GitHubToken, maxRetries = 3): GitHubToken {
  return withTokenProp(token, 'maxRetries', maxRetries)
}

/** Returns a copy of token with the user agent overridden. */
export function withUserAgent(token: GitHubToken, userAgent: string): GitHubToken {
  return withTokenProp(token, 'userAgent', userAgent)
}

/** Creates a signed GitHub App JWT. */
export function appJwt(appId: number, rsaKey: KeyObject): GitHubAppJwt {
  const header = Buffer.from(JSON.stringify({
    alg: 'RS256',
    typ: 'JWT',
  })).toString('base64url')

  const now = Math.floor(Date.now() / 1000)
  const payload = Buffer.from(JSON.stringify({
    iat: now - 60,
    exp: now + 60,
    iss: String(appId),
  })).toString('base64url')

  const sign = createSign('RSA-SHA256')
  sign.update(`${header}.${payload}`)
  const signature = sign.sign(rsaKey, 'base64url')

  return `${header}.${payload}.${signature}` as GitHubAppJwt
}

/** Get the GitHub App installation ID for repo. */
export async function getRepoInstallation(gh: GitHubApiUrl, jwt: GitHubAppJwt, repo: string): Promise<number> {
  const [resp, text] = await request(gh, jwt, 'GET', `repos/${repo}/installation`)
  if (resp.status !== 200) {
    throw new Error(jsonify`Response status ${resp.status} (body: ${text})`)
  }
  const obj = JSON.parse(text) as {
    id: number,
  }
  if (typeof obj?.id !== 'number') {
    throw new Error(jsonify`Response missing installation id`)
  }
  return obj.id
}

/** Create a GitHub App installation token for installId with contents:write permission on repo. */
export async function createInstallationToken(gh: GitHubApiUrl, jwt: GitHubAppJwt, repo: string, installId: number): Promise<GitHubInstallationToken> {
  const [resp, text] = await request(gh, jwt, 'POST', `app/installations/${installId}/access_tokens`, {
    repositories: [repo.replace(/^.+[/]/, '')],
    permissions: {
      contents: 'write',
    },
  })
  if (resp.status !== 201) {
    throw new Error(jsonify`Response status ${resp.status} (body: ${text})`)
  }
  const obj = JSON.parse(text) as {
    token: string,
    permissions: {
      contents: string,
    },
  }
  if (typeof obj?.token !== 'string') {
    throw new Error(jsonify`Response missing token`)
  }
  if (obj?.permissions?.contents !== 'write') {
    throw new Error('Installation does not have contents:write access')
  }
  return obj.token as GitHubInstallationToken
}

/** Revoke a GitHub App installation token. */
export async function revokeInstallationToken(gh: GitHubApiUrl, token: GitHubInstallationToken): Promise<void> {
  const [resp, text] = await request(gh, token, 'DELETE', 'installation/token')
  if (resp.status !== 204) {
    throw new Error(jsonify`Response status ${resp.status} (body: ${text})`)
  }
}

/** Make a GitHub REST API request. */
async function request(gh: GitHubApiUrl, token: GitHubToken, method: string, path: string, body?: any): Promise<[Response, string]> {
  const url = new URL(gh)
  if (!url.pathname.endsWith('/')) {
    url.pathname += '/'
  }
  url.pathname += path

  const headers = new Headers({
    'Accept': 'application/vnd.github+json',
    'X-GitHub-Api-Version': '2026-03-10',
  })
  if (body !== undefined) {
    headers.set('Content-Type', 'application/json')
    body = JSON.stringify(body)
  }

  try {
    return await fetchRetry(token, url, {
      method,
      headers,
      body,
    })
  } catch (err) {
    throw new Error(`${method} ${url}: ${err}`)
  }
}

export type GitObjectID = OID

export type Base64String = string & { __base64: true }

export interface CreateCommitOnBranchInput {
  branch: CommittableBranch
  expectedHeadOid: GitObjectID
  message: CommitMessage
  fileChanges: FileChanges
}

export interface CommittableBranch {
  repositoryNameWithOwner: string
  branchName: string
}

export interface CommitMessage {
  headline: string
  body: string
}

export interface FileChanges {
  additions: FileAddition[]
  deletions: FileDeletion[]
}

export interface FileAddition {
  contents: Base64String
  path: string
}

export interface FileDeletion {
  path: string
}

export function encodeBase64(buf: Buffer): Base64String {
  return buf.toString('base64') as Base64String
}

const commitThrottle = throttle(1000) // according to the GitHub recommendations for throttling content-generating requests

/** Create a commit using the GitHub GraphQL API. */
export async function createCommitOnBranch(gh: GitHubGraphqlUrl, token: GitHubToken, input: CreateCommitOnBranchInput): Promise<GitObjectID> {
  const url = new URL(gh)
  const method = 'POST'
  const headers = new Headers({
    'Accept': 'application/json',
    'Content-Type': 'application/json',
  })
  const body = JSON.stringify({
    query: `
      mutation($input: CreateCommitOnBranchInput!) {
        createCommitOnBranch(input: $input) {
          commit {
            oid
          }
        }
      }
    `,
    variables: {
      input,
    },
  })

  await commitThrottle()

  const [resp, text] = await fetchRetry(token, url, {
    method,
    headers,
    body,
  })

  const mt = resp.headers.get('Content-Type') ?? ''
  if (!mt.startsWith('application/json')) {
    if (resp.status !== 200) {
      throw new Error(jsonify`Response status ${resp.status} (body: ${text})`)
    }
    throw new Error(jsonify`Incorrect response type ${mt}`)
  }

  const obj = JSON.parse(text) as {
    errors?: {
      type: string,
      message: string,
    }[],
    data?: {
      createCommitOnBranch: {
        commit: {
          oid: GitObjectID,
        },
      },
    },
  }

  if (obj?.errors?.length) {
    for (const err of obj.errors) {
      if (err?.message?.includes('No commit exists with specified expectedHeadOid')) {
        throw new Error(`Remote branch head is behind local parent commit: ${err.type}: ${err.message}`)
      }
      if (err?.message?.includes('Expected branch to point to')) {
        throw new Error(`Local parent commit is behind remote branch head: ${err.type}: ${err.message}`)
      }
    }
    const msg = obj.errors.map(e => `\t${e.type}: ${e.message}`).join('\n')
    throw new Error(`GitHub GraphQL mutation failed:\n${msg}`)
  }
  if (!obj?.data?.createCommitOnBranch?.commit?.oid) {
    throw new Error(jsonify`GitHub created the commit but didn't return the oid`)
  }
  return obj?.data?.createCommitOnBranch?.commit?.oid
}

export interface RetryOptions {
  maxRetries?: number,
}

let retryLog: ((msg: string) => void) | undefined

export function setRetryLog(fn: (msg: string) => void): void {
  retryLog = fn
}

/**
 * Like fetch, but retries GitHub API requests using similar logic to
 * @octokit/plugin-throttling and @octokit/plugin-retry.
 */
export async function fetchRetry(token: GitHubToken, url: URL, init: RequestInit): Promise<[Response, string]> {
  const maxRetries = tokenProp(token, 'maxRetries') ?? 3 // like @octokit/plugin-retry
  const userAgent = tokenProp(token, 'userAgent') ?? globalUserAgent
  const doNotRetry = new Set([400, 401, 403, 404, 410, 422, 451]) // like @octokit/plugin-retry

  const headers = new Headers(init.headers)
  if (token) {
    headers.set('Authorization', `Bearer ${token}`)
  }
  if (userAgent) {
    headers.set('User-Agent', userAgent)
  }
  init = {...init, headers}

  for (let attempt = 1; ; attempt++) {
    const prefix = `${init.method ?? 'GET'} ${url} (attempt ${attempt})`
    debug(prefix)
    let resp: Response | undefined
    let isRetryable = false
    let isSecondaryRateLimit = false
    try {
      resp = await fetch(url, init)
      const text = await resp.text()

      if (resp) {
        debug(`${prefix}: status=${resp.status} ${Array.from(resp.headers.entries()).filter(([name]) => name.toLowerCase().startsWith('x-ratelimit-')).map(([name, value]) => `${name}=${JSON.stringify(value)}`).join(' ')}`)
      }

      if (resp.status >= 400) {
        if (text.includes('secondary rate')) { // like @octokit/plugin-throttling
          isRetryable = true
          isSecondaryRateLimit = true
          throw new Error(jsonify`hit secondary rate limit (body: ${text})`)
        }
        if (!doNotRetry.has(resp.status)) { // like @octokit/plugin-retry
          isRetryable = true
        }
        if (resp.headers.get('x-ratelimit-remaining') == '0' && (resp.status == 403 || resp.status == 429)) {
          throw new Error(jsonify`hit rate limit (body: ${text})`)
        }
        throw new Error((isRetryable ? `` : `non-retryable `) + jsonify`reponse status ${resp.status} (body: ${text})`)
      }

      return [resp, text] as const
    } catch (err) {
      debug(`${prefix}: failed (isRetryable=${isRetryable} isSecondaryRateLimit=${isSecondaryRateLimit}): ${err}`)

      if (!isRetryable || attempt > maxRetries) {
        throw new Error(`${prefix}: ${err instanceof Error ? err.message : err}`)
      }

      const retryAfter = resp?.headers.get('Retry-After')
      if (retryAfter) {
        let delay = (/^\d+$/.test(retryAfter)
          ? Math.max(0, parseInt(retryAfter, 10) * 1000)
          : Math.max(0, new Date(retryAfter).getTime() - Date.now()) + 1000)
        retryLog?.(`${prefix}: retrying failed (${err}) request after ${delay}ms (retry-after)`)
        debug(`${prefix}: retrying in ${delay}ms (retry-after)`)
        await new Promise(resolve => setTimeout(resolve, delay))
        continue
      }

      const rateLimitReset = resp?.headers.get('x-ratelimit-reset')
      if (rateLimitReset) {
        const delay = Math.max(0, parseInt(rateLimitReset) * 1000 - Date.now())
        retryLog?.(`${prefix}: retrying failed (${err}) request after ${delay}ms (x-ratelimit-reset)`)
        debug(`${prefix}: retrying in ${delay}ms (x-ratelimit-reset)`)
        await new Promise(resolve => setTimeout(resolve, delay))
        continue
      }

      const delay = isSecondaryRateLimit ? 60000 : (attempt ** 2) * 1000
      retryLog?.(`${prefix}: retrying failed (${err}) request after ${delay}ms`)
      debug(`${prefix}: retrying in ${delay}ms`)
      await new Promise(resolve => setTimeout(resolve, delay))
      continue
    }
  }
}

/**
 * Returns a promise which resolves at most once every interval.
 */
function throttle(interval: number): () => Promise<void> {
  let last: number | undefined
  return async () => {
    if (last) {
      const elapsed = Date.now() - last
      if (elapsed < interval) {
        await new Promise(resolve => setTimeout(resolve, interval - elapsed))
      }
    }
    last = Date.now()
  }
}

const symbols = {
  maxRetries: Symbol('maxRetries'),
  userAgent: Symbol('userAgent')
} as const satisfies { [K in keyof TokenProps]: symbol}

function withTokenProp<T extends GitHubToken, U extends TokenProp>(token: T, key: U, val: TokenProps[U]): TokenWithProp<U> {
  const tmp = new String(token)
  const sym = symbols[key]
  if (!sym) {
    throw new TypeError(jsonify`Unknown prop ${key}`)
  }
  for (const x of Object.values(symbols)) {
    if (Object.hasOwn(token, x) && x !== sym) {
      Object.defineProperty(tmp, x, {
        value: (token as any)[x],
        configurable: false,
        enumerable: false,
      })
    }
  }
  Object.defineProperty(tmp, sym, {
    value: val,
    configurable: false,
    enumerable: false,
  })
  return tmp as any
}

function tokenProp<T extends GitHubToken, U extends TokenProp>(token: T, key: U): T extends TokenWithProp<U> ? TokenProps[U] : TokenProps[U] | undefined {
  const sym = symbols[key]
  if (Object.hasOwn(token, sym)) {
    return (token as any)[sym] as TokenProps[U]
  }
  return undefined as any
}

export const __test = {
  withTokenProp,
  tokenProp,
}
